src/processors/default/ipfsMediaUploader.ts

```diff
@@ -0,0 +1,51 @@
+
+import { urlSource } from 'ipfs-http-client';
+import _ from 'lodash'
+
+import { Table } from '../../db/db';
+import { artworkNotOnIpfs, audioNotOnIpfs } from '../../triggers/ipfs';
+import { Clients, Processor } from '../../types/processor';
+import { ProcessedTrack } from '../../types/track';
+
+function processorFunction(sourceField: 'lossyAudioURL' | 'lossyArtworkURL', replaceField: 'lossyAudioIPFSHash' | 'lossyArtworkIPFSHash') {
+  return async (newTriggerItems: any, clients: Clients) => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const tracksByFilePath = _.keyBy(newTriggerItems, (track: ProcessedTrack) => new URL(track[sourceField]).pathname.split('/').pop());
+      
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const sources = newTriggerItems.map((track: ProcessedTrack) => {return urlSource(track[sourceField]!)} );
+    const allFiles = await clients.ipfs.client.addAll(sources, {
```

Comment:
hm we await for all here? so if 1 of 100 of the source urls is broken/massive/inaccessible, it will block all the others? i think ideally we should allow for their concurrency to work in parallel such that if a few urls time out, the rest are still processed properly?

Reply:
if we change the flow to handle download->upload->update to happen for a single track all in a row instead then you then also won't need this weird hack where you use the filePath to link a completed file download to a track (the filePath hack seems like it could have some edge cases where it breaks too, imi not sure?)

Reply:
yeah probably another reason to use `.add` instead of `.addAll` and do it manually, and insert any errors into the `nftProcessErrors` table

Reply:
To clarify what I mean here is i imagined something like.

```
tracks = [0,1,2,...100]

processTrack = track => { file = await ipfsAddTrack(track); updates.push({ ... file ...)}

promises = await tracks.map(processTrack)

await Promise.all(promises)

db update
```

Reply:
so rather than uploading the files using `urlSource` we should download each file and then upload it from our local copy?

Reply:
i'm not sure how urlSource works? i assumed it does dl the file and upload it under the hood, but i guess maybe it just tells the ipfs node to actually dl the file? either way, i think using urlSource as you do makes sense/is better as long as it's parallellized as we've discussed. so ignore my comment on downloading the file if it's not relevant.